### PR TITLE
New TimeZone and DateTimeWithTz VO

### PR DESCRIPTION
**Pull request by ajgarlag**
New classes to represent time zones and datetime with time zone info.

See #28

### CODE REVIEW

File: src/DateTime/DateTimeWithTz.php [hunk 1]
```diff
@@ -0,0 +1,156 @@
+<?php
+
+namespace ValueObjects\DateTime;
+
+use ValueObjects\Util\Util;
+use ValueObjects\ValueObjectInterface;
+
+class DateTimeWithTz implements ValueObjectInterface
```
Comment (nicolopignatelli) [inline src/DateTime/DateTimeWithTz.php]: can you please rename to DateTimeWithTimezone?

File: src/DateTime/DateTimeWithTz.php [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+{
+    /** @var DateTime */
+    protected $datetime;
```
Comment (nicolopignatelli) [inline src/DateTime/DateTimeWithTz.php]: please rename to $dateTime

File: src/DateTime/DateTimeWithTz.php [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+    /** @var TimeZone */
+    protected $timezone;
```
Comment (nicolopignatelli) [inline src/DateTime/DateTimeWithTz.php]: please rename to $timeZone

File: src/DateTime/DateTimeWithTz.php [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    public static function fromNativeDateTime(\DateTime $nativeDatetime)
+    {
+        $datetime = DateTime::fromNativeDateTime($nativeDatetime);
+        $timezone = TimeZone::fromNativeDateTimeZone($nativeDatetime->getTimezone());
+
+        return new static($datetime, $timezone);
+    }
+
+    /**
+     * Returns current DateTimeWithTz
```
Comment (nicolopignatelli) [inline src/DateTime/DateTimeWithTz.php]: "Returns a DateTimeWithTimeZone object using current DateTime and default TimeZone"

File: src/DateTime/DateTimeWithTz.php [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+     *
+     * @return DateTimeWithTz
+     */
+    public static function now()
+    {
+        return new static(DateTime::now(), TimeZone::defaultTz());
+    }
+
+    /**
+     * Returns a new DateTime object
```
Comment (nicolopignatelli) [inline src/DateTime/DateTimeWithTz.php]: Returns a new DateTimeWithTimeZone object

File: src/DateTime/DateTimeWithTz.php [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+     *
+     * @parm DateTime $datetime
```
Comment (nicolopignatelli) [inline src/DateTime/DateTimeWithTz.php]: @param

File: src/DateTime/DateTimeWithTz.php [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+     * @param  ValueObjectInterface $dateTimeWithTz
+     * @return bool
+     */
+    public function sameValueAs(ValueObjectInterface $dateTimeWithTz)
+    {
+        if (false === Util::classEquals($this, $dateTimeWithTz)) {
+            return false;
+        }
+
+        return $this->getDateTime()->sameValueAs($dateTimeWithTz->getDateTime()) && $this->getTimeZone()->sameValueAs($dateTimeWithTz->getTimeZone());
```
Comment (nicolopignatelli) [inline src/DateTime/DateTimeWithTz.php]: please add a line break between after the && for shorter lines (PSR-2 compliant)

File: src/DateTime/DateTimeWithTz.php [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        return $this->timezone;
+    }
+
+    /**
+     * Returns a native PHP \DateTime version of the current DateTimeWithTz.
+     *
+     * @return \DateTime
+     */
+    public function toNativeDateTime()
+    {
```
Comment (nicolopignatelli) [inline src/DateTime/DateTimeWithTz.php]: please define $date = $this->getDateTime()->getDate() and $time = $this->getDateTime()->getTime()
so you don't have to repeat them for each assigned var

File: src/DateTime/TimeZone.php [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        return static::fromNative($timezone->getName());
+    }
+
+    /**
+     * Returns default TimeZone
+     *
+     * @return self
+     */
+    public static function defaultTz()
```
Comment (nicolopignatelli) [inline src/DateTime/TimeZone.php]: I think that just TimeZone::default() is enough. TimeZone::defaultTz() is redundant

File: tests/DateTime/DateTimeWithTzTest.php [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+use ValueObjects\DateTime\TimeZone;
+use ValueObjects\String\String;
+
+class DateTimeWithTzTest extends TestCase
+{
+    public function testFromNative()
+    {
+        $fromNativeDateTimeWithTz  = DateTimeWithTz::fromNative(2013, 'December', 21, 10, 20, 34, 'Europe/Madrid');
+        $constructedDateTimeWithTz = new DateTimeWithTz(
+                                        new DateTime(
```
Comment (nicolopignatelli) [inline tests/DateTime/DateTimeWithTzTest.php]: can you respect the default progressive indentation?

File: tests/DateTime/TimeZoneTest.php [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+        $this->assertEquals('Europe/Madrid', $timeZone->__toString());
+    }
+
+    /**
+     * @expectedException \ValueObjects\DateTime\Exception\InvalidTimeZoneException
+     */
+    public function testExceptionOnInvalidTimeZoneName()
+    {
+        $timeZone = new TimeZone(new String('Mars/Phobos'));
```
Comment (nicolopignatelli) [inline tests/DateTime/TimeZoneTest.php]: lol :laughing: best timezone ever!

**Comment by nicolopignatelli** [comment]
Code review done, can you please fix and rebase?

Great job!

File: src/DateTime/TimeZone.php (hunk 9 above)
Comment (ajgarlag) [inline src/DateTime/TimeZone.php]: `default` is a reserved word

Comment (ajgarlag) [inline src/DateTime/TimeZone.php]: Maybe, `TimeZone::byDefault()` is a better name. What do you think?

Comment (nicolopignatelli) [inline src/DateTime/TimeZone.php]: TimeZone::byDefault(), TimeZone::fromDefault(), TimeZone::getDefault()

You choose :)

**Comment by nicolopignatelli** [comment]
Seems good to me. Please squash all commits into one, then we can merge

**Comment by ajgarlag** [comment]
Ready to merge

**Comment by nicolopignatelli** [comment]
Great job, ty!

**Pull request merged**